spark/v3.3/spark/src/test/java/org/apache/iceberg/spark/TestFunctionCatalog.java

```diff
@@ -0,0 +1,165 @@
+import org.apache.spark.sql.types.StructType;
+import org.assertj.core.api.Assertions;
+import org.junit.After;
+import org.junit.Assert;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.Parameterized;
+
+@RunWith(Parameterized.class)
+public class TestFunctionCatalog extends SparkCatalogTestBase {
+  @Parameterized.Parameters(name = "catalogConfig = {0}")
+  public static Object[][] parameters() {
+    return new Object[][] {
+      {SparkCatalogConfig.HADOOP}, {SparkCatalogConfig.HIVE}, {SparkCatalogConfig.SPARK}
+    };
+  }
+
+  private static final Namespace NS = Namespace.of("db");
+  private final boolean isSessionCatalog;
+  private final String fullNamespace;
+  private final FunctionCatalog asFunctionCatalog;
+
+  public TestFunctionCatalog(SparkCatalogConfig catalogConfig) {
+    super(catalogConfig);
+    this.isSessionCatalog = "spark_catalog".equals(catalogName);
+    this.fullNamespace = (isSessionCatalog ? "" : catalogName + ".") + NS;
+    this.asFunctionCatalog = castToFunctionCatalog(catalogName);
+  }
+
+  @Before
+  public void createNamespace() {
+    sql("CREATE NAMESPACE IF NOT EXISTS %s", fullNamespace);
+  }
+
+  @After
+  public void cleanNamespaces() {
+    sql("DROP NAMESPACE IF EXISTS %s", fullNamespace);
+  }
+
+  @Test
+  public void testLoadListAndUseFunctionsFromSystemNamespace()
+      throws NoSuchFunctionException, NoSuchNamespaceException {
+    // Session catalog requires that the namespace actually exists
+    if (isSessionCatalog) {
+      sql("CREATE NAMESPACE IF NOT EXISTS system");
+    }
+
+    try {
+      String[] namespace = {"system"};
+      String name = "iceberg_version";
+      Identifier identifier = Identifier.of(namespace, name);
+
+      assertListingLoadingAndBindingFrom(identifier);
+    } finally {
+      if (isSessionCatalog) {
+        sql("DROP NAMESPACE IF EXISTS system");
+      }
+    }
+  }
+
+  @Test
+  public void testLoadListAndUseFunctionsFromEmptyNamespace()
+      throws NoSuchFunctionException, NoSuchNamespaceException {
+    String[] namespace = {};
+    String name = "iceberg_version";
+    Identifier identifier = Identifier.of(namespace, name);
+
+    assertListingLoadingAndBindingFrom(identifier);
+  }
+
+  @Test
+  public void testCannotLoadFunctionsFromInvalidNamespace() {
+    AssertHelpers.assertThrows(
+        "Function Catalog functions should only be accessible from the system namespace and empty namespace",
+        AnalysisException.class,
+        "Undefined function",
+        () -> sql("SELECT %s.iceberg_version()", fullNamespace));
+  }
+
+  @Test
+  public void testCannotUseUndefinedFunction() {
+    AssertHelpers.assertThrows(
+        "Using an undefined function should throw",
+        AnalysisException.class,
+        "Undefined function",
+        () -> sql("SELECT undefined_function(1, 2)"));
+  }
+
+  private void assertListingLoadingAndBindingFrom(Identifier identifier)
```

Comment:
What is the value of testing all of these things directly? I would expect these tests to run SQL, like `SELECT iceberg_version()` and verify the result.

Think about what you need to validate in generic catalog-level tests. You need to validate that you can load and call a function, assuming that it works properly. You don't need to test the function itself (unbound to bound, for example) because that isn't the goal of this suite.

I think that this suite should verify, using SQL, that you can call the `iceberg_version` method and it shows up in listings. You can also verify that it is in both `system` and the empty namespace. Beyond that, I don't think this needs to do much.

Reply:
This is partially because of oddities between what can be done with V2 functions in the session catalog and elsewhere (non-session catalogs).

I addded the tests in these styles because this is how the storage partitioned joins function resolution actually happens. The function is created with a "fake" input Schema to addd `numBuckets` and then called through this pass.

And then not testing certain other things through plain SQL is because of limitations / differences in what can be done in the session catalog vs elsewhere. Per the Spark 3.3 tests for example, only the session catalog can use describe (and my own findings).

I'll try to refactor to be more in line with your thought though as I don't disagree 👍

Reply:
In Spark's `DataSourceV2FunctionSuite` for example, they have the following test (which is similar to the results I get) https://github.com/apache/spark/blob/144d4c546f7023b20e07619134feca1a46017a5f/sql/core/src/test/scala/org/apache/spark/sql/connector/DataSourceV2FunctionSuite.scala#L157-L169

But I'll refactor to be more like that suite and what you mentioned. I don't disagree by any means with what you say.

Reply:
Ok so we can't exactly test `DESCRIBE FUNCTION` as mentioned. We might want to put in the work upstream to fix that for V2 functions.

When testing `DESCRIBE FUNCTION` with the session catalog (from SQL), it resolves to the v1 session catalog and I get the following error

Reply:
```
scalarSQL("DESCRIBE FUNCTION iceberg_version");
```

class org.apache.iceberg.spark.functions.IcebergVersionFunction cannot be cast to class org.apache.spark.sql.internal.connector.V1Function (org.apache.iceberg.spark.functions.IcebergVersionFunction and org.apache.spark.sql.internal.connector.V1Function are in unnamed module of loader 'app')
java.lang.ClassCastException: class org.apache.iceberg.spark.functions.IcebergVersionFunction cannot be cast to class org.apache.spark.sql.internal.connector.V1Function (org.apache.iceberg.spark.functions.IcebergVersionFunction and org.apache.spark.sql.internal.connector.V1Function are in unnamed module of loader 'app')
	at org.apache.spark.sql.catalyst.analysis.ResolveSessionCatalog$$anonfun$apply$1.applyOrElse(ResolveSessionCatalog.scala:389)
	at org.apache.spark.sql.catalyst.analysis.ResolveSessionCatalog$$anonfun$apply$1.applyOrElse(ResolveSessionCatalog.scala:49)
	at org.apache.spark.sql.catalyst.plans.logical.AnalysisHelper.$anonfun$resolveOperatorsUpWithPruning$3(AnalysisHelper.scala:138)

So it seems that we can't `DESCRIBE` on a non-v1 function from within the tests. The tests in Spark also support this idea.

So calling the function directly with code is the best it seems we can do for those, but the function resolution (e.e. `SELECT system.iceberg_version` could be moved over here and we could get rid of `TestIcebergVersionFunction`.